### PR TITLE
[GDBJIT] Generate more complete debug information

### PR DESCRIPTION
**Pull request by kvochko**
This PR adds support for classes, value types, strings, arrays and static variables to GDBJIT.
@janvorli @mikem8361 PTAL
CC @Dmitri-Botcharnikov @chunseoklee @seanshpark @lucenticus @mskvortsov @ayuckhulk

### CODE REVIEW

File: src/vm/gdbjit.cpp [hunk 1]
```diff
@@ -600,6 +673,157 @@ struct __attribute__((packed)) DebugInfoVar
...
+void BaseTypeInfo::DumpDebugInfo(char* ptr, int& offset)
+{
+    if (m_type_offset != 0)
+    {
+        return;
+    }
+
+    if (ptr != nullptr)
+    {
+        NewHolder<DebugInfoType> bufType = new (nothrow) DebugInfoType;
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: It seems the bufType can be a stack variable, it is small and you copy its contents out after you build it anyways.
And the same at several places below where you use the holder for other debug info stuff.

File: src/vm/gdbjit.cpp [hunk 2]
```diff
@@ -920,7 +1147,7 @@ void NotifyGdb::MethodCompiled(MethodDesc* MethodDescPtr)
 
     elfFile.MemPtr.SuppressRelease();
 
-#ifdef GDBJIT_DUMPELF
+#if 1//def GDBJIT_DUMPELF
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: This seems like a forgotten change made during debugging that should have been reverted.

File: src/vm/gdbjit.cpp [hunk 3]
```diff
@@ -1333,15 +1565,23 @@ int NotifyGdb::GetFrameLocation(int nativeOffset, char* bufVarLoc)
     return len + 2;  // We add '2' because first 2 bytes contain length of expression and DW_OP_fbreg operation.
 }
 /* Build tge DWARF .debug_info section */
-bool NotifyGdb::BuildDebugInfo(MemBuf& buf, NewArrayHolder<ArgsDebugInfo> &argsDebug, unsigned int argsDebugSize,
+bool NotifyGdb::BuildDebugInfo(MemBuf& buf, PMT_TypeInfoMap pTypeMap, NewArrayHolder<ArgsDebugInfo> &argsDebug, unsigned int argsDebugSize,
                                NewArrayHolder<LocalsDebugInfo> &localsDebug, unsigned int localsDebugSize)
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: Can you please pass in the parameter las LocalsDebugInfo* instead? It seems we don't need to pass reference to the whole holder and it puts an implementation detail to the method signature.

**Comment by kvochko** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test

File: src/vm/gdbjit.cpp [hunk 4]
```diff
@@ -371,7 +461,7 @@ GetDebugInfoFromPDB(MethodDesc* MethodDescPtr, SymbolsInfo** symInfo, unsigned i
     StackScratchBuffer scratch;
     const char* szModName = modName.GetUTF8(scratch);
 
-    MethodDebugInfo* methodDebugInfo = new (nothrow) MethodDebugInfo();
+    NewHolder<MethodDebugInfo> methodDebugInfo = new (nothrow) MethodDebugInfo();
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: This is yet another struct that can be allocated on stack instead of using holder.

**Comment by kvochko** [comment]
@dotnet-bot test Linux ARM Emulator Cross Debug Build please

File: src/vm/gdbjit.cpp [hunk 5]
```diff
@@ -1637,21 +2135,23 @@ bool NotifyGdb::BuildSectionTable(MemBuf& buf)
...
     buf.MemSize = sizeof(Elf_Shdr) * (SectionNamesCount - 1);
     return true;
 }
 
 /* Build the ELF header */
 bool NotifyGdb::BuildELFHeader(MemBuf& buf)
 {
-    Elf_Ehdr* header = new (nothrow) Elf_Ehdr;
-    buf.MemPtr = reinterpret_cast<char*>(header);
+    NewHolder<Elf_Ehdr> header = new (nothrow) Elf_Ehdr;
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: The holder here is not necessary - you always call SuppressRelease(), so it is better to use just a plain pointer.

**Comment by chunseoklee** [comment]
@kvochko This PR seems to have a bug. 
PR is rebased on 3d05ba06eb2e549e3501aaa11b04d2ef1af450b

When debugging with CORECLR_GDBJIT of multiple files, corerun terminates with seg fauilt as shown below :
(or course, without PR, breakpoint works fine.)
```

twoflower@js1-desktop ~/d/coreoverlay_x64> export CORECLR_GDBJIT=Application1.exe,ClassLibrary1.dll
twoflower@js1-desktop ~/d/coreoverlay_x64> lldb corerun -- Application1.exe
(lldb) target create "corerun"
Current executable set to 'corerun' (x86_64).
(lldb) settings set -- target.run-args  "Application1.exe"
(lldb) b Application1.cs:19
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) r
Process 7543 launched: '/home/twoflower/dev/coreoverlay_x64/corerun' (x86_64)
1 location added to breakpoint 1
libcoreclr.so was compiled with optimization - stepping may behave oddly; variables may not be available.
Process 7543 stopped
* thread #1: tid = 7543, 0x00007ffff6572949 libcoreclr.so`TypeHandle::GetTypeKey() const [inlined] MethodTable::GetFlag(flag=enum_flag_Category_ValueType_Mask) const at methodtable.h:4081, name = 'corerun', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
    frame #0: 0x00007ffff6572949 libcoreclr.so`TypeHandle::GetTypeKey() const [inlined] MethodTable::GetFlag(flag=enum_flag_Category_ValueType_Mask) const at methodtable.h:4081 [opt]
   4078     __forceinline DWORD GetFlag(WFLAGS_HIGH_ENUM flag) const
   4079     {
   4080         LIMITED_METHOD_DAC_CONTRACT;
-> 4081         return m_dwFlags & flag;
   4082     }
   4083     __forceinline BOOL TestFlagWithMask(WFLAGS_HIGH_ENUM mask, WFLAGS_HIGH_ENUM flag) const
   4084     {
(lldb)
```

```
using System;
using ClassLibrary1;

namespace Application1
{
    class Program
    {
        static int classint = 10;
        static void Show(int i)
        {
            int mainint = 100;
            string abc = "kasdjflaks";
            Console.WriteLine(i + "Show called!!!");
        }
        static void Main(string[] args)
        {
            MyLibrary mylib = new MyLibrary();
            mylib.MyMethod();            
            int mainint = 100;
            string abc = "kasdjflaks";
            for(int i=0;i<5;i++) 
                Show(i); 
            Console.WriteLine("Hello World!!!" + mainint + abc);
        }
    }
}


#### ClassLibrary1.cs
using System;

namespace ClassLibrary1
{
    public class MyLibrary
    {
        public string MyMethod()
        {
            return "Hello Library!!!";
        }
    }
}
    
```

**Comment by chunseoklee** [comment]
```
(lldb) bt
* thread #1: tid = 7543, 0x00007ffff6572949 libcoreclr.so`TypeHandle::GetTypeKey() const [inlined] MethodTable::GetFlag(flag=enum_flag_Category_ValueType_Mask) const at methodtable.h:4081, name = 'corerun', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x00007ffff6572949 libcoreclr.so`TypeHandle::GetTypeKey() const [inlined] MethodTable::GetFlag(flag=enum_flag_Category_ValueType_Mask) const at methodtable.h:4081 [opt]
    frame #1: 0x00007ffff6572949 libcoreclr.so`TypeHandle::GetTypeKey() const [inlined] MethodTable::IsArray() at methodtable.h:2854 [opt]
    frame #2: 0x00007ffff6572949 libcoreclr.so`TypeHandle::GetTypeKey(this=0x00007fffffffd100) const + 41 at typehandle.cpp:1930 [opt]
    frame #3: 0x00007ffff658eb29 libcoreclr.so`GetTypeInfoFromTypeHandle(typeHandle=<unavailable>, pTypeMap=0x00000000006fec30) + 57 at gdbjit.cpp:22 [opt]
    frame #4: 0x00007ffff65922dd libcoreclr.so`NotifyGdb::MethodCompiled(MethodDesc*) [inlined] GetArgTypeInfo(MethodDescPtr=0x00007fff7c824290, ilIndex=0) + 1629 at gdbjit.cpp:225 [opt]
    frame #5: 0x00007ffff65922c4 libcoreclr.so`NotifyGdb::MethodCompiled(MethodDescPtr=<unavailable>) + 1604 at gdbjit.cpp:1333 [opt]
    frame #6: 0x00007ffff657c14f libcoreclr.so`DACNotify::DoJITNotification(MethodDescPtr=0x00007fff7c824290) + 79 at util.cpp:3425 [opt]
    frame #7: 0x00007ffff6705a43 libcoreclr.so`MethodDesc::MakeJitWorker(COR_ILMETHOD_DECODER*, unsigned int, unsigned int) [inlined] DACNotifyCompilationFinished(MethodDesc*) + 29 at prestub.cpp:234 [opt]
    frame #8: 0x00007ffff6705a26 libcoreclr.so`MethodDesc::MakeJitWorker(this=0x00007fff7c824290, ILHeader=0x00000000006a0280, flags=0, flags2=0) + 1238 at prestub.cpp:633 [opt]
    frame #9: 0x00007ffff670714d libcoreclr.so`MethodDesc::DoPrestub(this=0x00007fff7c824290, pDispatchingMT=0x0000000000000000) + 1405 at prestub.cpp:1458 [opt]
    frame #10: 0x00007ffff6706973 libcoreclr.so`::PreStubWorker(pTransitionBlock=<unavailable>, pMD=0x00007fff7c824290) + 403 at prestub.cpp:1045 [opt]
    frame #11: 0x00007ffff66896b8 libcoreclr.so`ThePreStub + 92
    frame #12: 0x00007fff7d2c0cc7 JIT(0x6ff880)`Main(args=0x00007fff5801d380) + 103 at Application1.cs:18
    frame #13: 0x00007ffff6688cd3 libcoreclr.so`CallDescrWorkerInternal + 124
    frame #14: 0x00007ffff659d860 libcoreclr.so`MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) + 83 at callhelpers.cpp:89 [opt]
    frame #15: 0x00007ffff659d80d libcoreclr.so`MethodDescCallSite::CallTargetWorker(this=<unavailable>, pArguments=0x00007fffffffdbe8, pReturnValue=0x0000000000000000, cbReturnValue=0) + 1197 at callhelpers.cpp:656 [opt]
    frame #16: 0x00007ffff66aae7b libcoreclr.so`RunMain(MethodDesc*, short, int*, PtrArray**) [inlined] MethodDescCallSite::Call(this=<unavailable>, pArguments=<unavailable>) + 5 at callhelpers.h:433 [opt]
    frame #17: 0x00007ffff66aae76 libcoreclr.so`RunMain(MethodDesc*, short, int*, PtrArray**) + 363 at assembly.cpp:2637 [opt]
    frame #18: 0x00007ffff66aad0b libcoreclr.so`RunMain(MethodDesc*, short, int*, PtrArray**) + 87 at assembly.cpp:2655 [opt]
    frame #19: 0x00007ffff66aacb4 libcoreclr.so`RunMain(pFD=<unavailable>, numSkipArgs=<unavailable>, piRetVal=<unavailable>, stringArgs=<unavailable>) + 292 at assembly.cpp:2655 [opt]
    frame #20: 0x00007ffff66ab113 libcoreclr.so`Assembly::ExecuteMainMethod(this=0x00000000006a2260, stringArgs=0x00007fffffffe028, waitForOtherThreads=YES) + 227 at assembly.cpp:2762 [opt]
    frame #21: 0x00007ffff64d96c2 libcoreclr.so`CorHost2::ExecuteAssembly(this=<unavailable>, dwAppDomainId=<unavailable>, pwzAssemblyPath=<unavailable>, argc=<unavailable>, argv=<unavailable>, pReturnValue=<unavailable>) + 482 at corhost.cpp:1349 [opt]
    frame #22: 0x00007ffff64b4b24 libcoreclr.so`::coreclr_execute_assembly(hostHandle=<unavailable>, domainId=<unavailable>, argc=<unavailable>, argv=<unavailable>, managedAssemblyPath=<unavailable>, exitCode=<unavailable>) + 260 at unixinterface.cpp:375 [opt]
    frame #23: 0x0000000000402bd2 corerun`ExecuteManagedAssembly(currentExeAbsolutePath=<unavailable>, clrFilesAbsolutePath=<unavailable>, managedAssemblyAbsolutePath=<unavailable>, managedAssemblyArgc=<unavailable>, managedAssemblyArgv=<unavailable>) + 882 at coreruncommon.cpp:419 [opt]
    frame #24: 0x0000000000401cc5 corerun`corerun(argc=<unavailable>, argv=0x00007fffffffe408) + 293 at corerun.cpp:149 [opt]
    frame #25: 0x00007ffff6df4f45 libc.so.6`__libc_start_main(main=(corerun`main at corerun.cpp:161), argc=2, argv=0x00007fffffffe408, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007fffffffe3f8) + 245 at libc-start.c:287
    frame #26: 0x0000000000401929 corerun

```

**Comment by lucenticus** [comment]
CC @Buyduck

**Comment by kvochko** [comment]
@chunseoklee thank you for testing this! Your example works for me with the latest fix, could you please verify?

**Comment by janvorli** [comment]
@kvochko I can see you have added a commit fixing a problem. Are you working on more fixes or is it ready to go in?

**Comment by chunseoklee** [comment]
@kvochko it works now.

**Comment by chunseoklee** [comment]
@kvochko I have noticed another issue on this PR : step-in(-exec-step) does not work with this PR.

**Comment by ayuckhulk** [comment]
Step in is going to be fixed in upcoming PR, but it depends on changes in current PR.

**Comment by kvochko** [comment]
@janvorli we're not working on any more fixes right now, it's ready to go in!

**Pull request merged**